File: bowling-game/bowling-game-cs/BowlingGame/Game.cs
namespace BowlingGame;

public class Game
{
    private readonly int[] rolls = new int[21];
    private int currentRoll = 0;
    
    public void Roll(int pins)
    {
        rolls[currentRoll++] = pins;
    }

    public int Score()
    {
        int score = 0;
        int firstInFrame = 0;
        for (int frame = 0; frame < 10; frame++)
        {
            if (IsStrike(firstInFrame))
            {
                score += 10 + NextTwoBallsForStrike(firstInFrame);
                firstInFrame++;
            }
            else if (IsSpare(firstInFrame))
            {
                score += 10 + NextBallForSpare(firstInFrame);
                firstInFrame += 2;
            }
            else
            {
                score += TwoBallsInFrame(firstInFrame);
                firstInFrame += 2;
            }
        }
        return score;
    }

    private int NextTwoBallsForStrike(int firstInFrame)
    {
        return rolls[firstInFrame + 1] + rolls[firstInFrame + 2];
    }

    private int NextBallForSpare(int firstInFrame)
    {
        return rolls[firstInFrame + 1];
    }

    private int TwoBallsInFrame(int firstInFrame)
    {
        return rolls[firstInFrame] + rolls[firstInFrame + 1];
    }

    private bool IsStrike(int firstInFrame)
    {
        return rolls[firstInFrame] == 10;
    }

    private bool IsSpare(int firstInFrame)
    {
        return rolls[firstInFrame] + rolls[firstInFrame + 1] == 10;
    }
}

File: bowling-game/bowling-game-py/03/Game.py
class Game:
    def __init__(self):
        self.rolls = []

    def roll(self, pins) -> None:
        self.rolls.append(pins)

    def score(self) -> int:
        score = 0
        first_in_frame = 0
        for frame in range(10):
            score += self.two_balls_in_frame(first_in_frame)
            first_in_frame += 2
        return score

    def two_balls_in_frame(self, first_in_frame) -> int:
        return self.rolls[first_in_frame] + self.rolls[first_in_frame + 1]